File: tools/vf-evolve/tasks/llm_sql_cache/initial.py
"""
LLM SQL Cache - Initial Seed (LRU Cache with Prefix Matching)
ShinkaEvolve will evolve this toward optimal caching/retrieval policies.

Simple LRU cache that stores SQL query results keyed by normalized query text,
with prefix-based lookup for similar queries.
"""

from collections import OrderedDict
from dataclasses import dataclass, field
from typing import Any, Dict, List, Optional, Tuple
import hashlib
import re


@dataclass
class CacheEntry:
    """A cached SQL query result."""
    query: str
    normalized: str
    result: Any
    hit_count: int = 0
    creation_time: float = 0.0


@dataclass
class CacheStats:
    """Cache performance statistics."""
    hits: int = 0
    misses: int = 0
    prefix_hits: int = 0
    evictions: int = 0
    size: int = 0

    @property
    def hit_rate(self) -> float:
        total = self.hits + self.misses
        return self.hits / total if total > 0 else 0.0


class LruSqlCache:
    """LRU cache with prefix matching for SQL queries."""

    def __init__(self, capacity: int = 1000):
        assert capacity > 0, "capacity must be positive"
        self.capacity = capacity
        self.cache: OrderedDict[str, CacheEntry] = OrderedDict()
        self.prefix_index: Dict[str, List[str]] = {}  # prefix -> [cache_keys]
        self.stats = CacheStats()

    @staticmethod
    def normalize_query(query: str) -> str:
        """Normalize SQL query for cache key generation."""
        q = query.strip().lower()
        # Collapse whitespace
        q = re.sub(r'\s+', ' ', q)
        # Remove trailing semicolons
        q = q.rstrip(';')
        # Normalize literal values to placeholders for prefix matching
        q = re.sub(r"'[^']*'", "'?'", q)
        q = re.sub(r'\b\d+\b', '?', q)
        return q

    @staticmethod
    def _extract_prefix(normalized: str, depth: int = 3) -> str:
        """Extract a prefix from normalized query (first N tokens)."""
        tokens = normalized.split()
        prefix_tokens = tokens[:depth]
        return ' '.join(prefix_tokens)

    def _cache_key(self, normalized: str) -> str:
        """Generate a cache key from normalized query."""
        return hashlib.md5(normalized.encode()).hexdigest()

    def get(self, query: str) -> Optional[Any]:
        """Look up a query in cache. Returns result or None."""
        normalized = self.normalize_query(query)
        key = self._cache_key(normalized)

        if key in self.cache:
            # Move to end (most recently used)
            self.cache.move_to_end(key)
            entry = self.cache[key]
            entry.hit_count += 1
            self.stats.hits += 1
            return entry.result

        self.stats.misses += 1
        return None

    def get_prefix_match(self, query: str) -> Optional[Tuple[str, Any]]:
        """Find a cached result with matching prefix (approximate match).
        Returns (original_query, result) or None.
        """
        normalized = self.normalize_query(query)
        prefix = self._extract_prefix(normalized)

        if prefix in self.prefix_index:
            for key in self.prefix_index[prefix]:
                if key in self.cache:
                    self.cache.move_to_end(key)
                    entry = self.cache[key]
                    entry.hit_count += 1
                    self.stats.prefix_hits += 1
                    return (entry.query, entry.result)

        return None

    def put(self, query: str, result: Any, timestamp: float = 0.0) -> None:
        """Insert a query result into the cache."""
        normalized = self.normalize_query(query)
        key = self._cache_key(normalized)
        prefix = self._extract_prefix(normalized)

        # Evict if at capacity
        if key not in self.cache and len(self.cache) >= self.capacity:
            self._evict()

        # Insert or update
        if key in self.cache:
            self.cache.move_to_end(key)
            self.cache[key].result = result
        else:
            entry = CacheEntry(
                query=query,
                normalized=normalized,
                result=result,
                creation_time=timestamp,
            )
            self.cache[key] = entry

            # Update prefix index
            if prefix not in self.prefix_index:
                self.prefix_index[prefix] = []
            self.prefix_index[prefix].append(key)

        self.stats.size = len(self.cache)

    def _evict(self) -> None:
        """Evict the least recently used entry."""
        if not self.cache:
            return
        key, entry = self.cache.popitem(last=False)  # Remove oldest
        prefix = self._extract_prefix(entry.normalized)
        if prefix in self.prefix_index:
            self.prefix_index[prefix] = [
                k for k in self.prefix_index[prefix] if k != key
            ]
            if not self.prefix_index[prefix]:
                del self.prefix_index[prefix]
        self.stats.evictions += 1
        self.stats.size = len(self.cache)

    def clear(self) -> None:
        """Clear all cached entries."""
        self.cache.clear()
        self.prefix_index.clear()
        self.stats.size = 0

    def get_stats(self) -> CacheStats:
        """Return current cache statistics."""
        return self.stats


def cache_strategy(queries, memory_budget, distinct_prefixes):
    """Adapter for ShinkaEvolve simulator interface.

    Args:
        queries: list of query strings.
        memory_budget: memory budget in GB.
        distinct_prefixes: number of distinct prefixes.

    Returns:
        dict with hit_rate.
    """
    cache = LruSqlCache(capacity=min(distinct_prefixes, memory_budget * 100))
    hits = 0
    for i, query in enumerate(queries):
        if cache.get(query) is not None:
            hits += 1
        else:
            cache.put(query, i)
    return {"hit_rate": hits / len(queries) if queries else 0.0}


if __name__ == "__main__":
    cache = LruSqlCache(capacity=5)

    # Store some queries
    cache.put("SELECT count(*) FROM logs WHERE service='nginx'", 42)
    cache.put("SELECT count(*) FROM logs WHERE service='redis'", 17)
    cache.put("SELECT avg(duration) FROM spans WHERE service='api'", 0.23)

    # Exact hit
    result = cache.get("SELECT count(*) FROM logs WHERE service='nginx'")
    print(f"Exact hit: {result}")  # 42

    # Miss
    result = cache.get("SELECT max(duration) FROM spans")
    print(f"Miss: {result}")  # None

    # Prefix match (different literal value but same structure)
    match = cache.get_prefix_match("SELECT count(*) FROM logs WHERE env='prod'")
    if match:
        print(f"Prefix match: query='{match[0]}' result={match[1]}")

    stats = cache.get_stats()
    print(f"\nStats: hits={stats.hits}, misses={stats.misses}, "
          f"prefix_hits={stats.prefix_hits}, hit_rate={stats.hit_rate:.2%}")

File: tools/vf-evolve/tasks/llm_sql_cache/initial.rs
/// LLM SQL Cache - Initial Seed (LRU Cache with Prefix Matching)
/// ShinkaEvolve will evolve this toward optimal caching/retrieval policies.
///
/// Simple LRU cache keyed by normalized query text, with prefix lookup
/// for structurally similar queries.

use std::collections::HashMap;

/// Cache performance statistics.
#[derive(Debug, Clone, Default)]
pub struct CacheStats {
    pub hits: u64,
    pub misses: u64,
    pub prefix_hits: u64,
    pub evictions: u64,
    pub size: u64,
}

impl CacheStats {
    pub fn hit_rate(&self) -> f64 {
        let total = self.hits + self.misses;
        if total > 0 {
            self.hits as f64 / total as f64
        } else {
            0.0
        }
    }
}

/// A cached SQL query result.
#[derive(Debug, Clone)]
struct CacheEntry {
    query: String,
    normalized: String,
    result_id: u64,
    hit_count: u64,
}

/// LRU cache for SQL queries with prefix-based approximate matching.
pub struct LruSqlCache {
    capacity: u64,
    /// Ordered map: most recently used at end. Key = normalized query hash.
    entries: Vec<(u64, CacheEntry)>,
    /// Prefix -> list of entry keys for approximate matching.
    prefix_index: HashMap<String, Vec<u64>>,
    stats: CacheStats,
    next_key: u64,
}

/// Normalize a SQL query for cache key generation.
pub fn normalize_query(query: &str) -> String {
    let mut q = query.trim().to_lowercase();
    // Collapse whitespace
    let mut prev_space = false;
    q = q
        .chars()
        .filter(|&c| {
            if c.is_whitespace() {
                if prev_space {
                    return false;
                }
                prev_space = true;
            } else {
                prev_space = false;
            }
            true
        })
        .collect();
    // Remove trailing semicolons
    while q.ends_with(';') {
        q.pop();
    }
    q
}

fn extract_prefix(normalized: &str, depth: usize) -> String {
    normalized
        .split_whitespace()
        .take(depth)
        .collect::<Vec<&str>>()
        .join(" ")
}

fn simple_hash(s: &str) -> u64 {
    let mut h: u64 = 0xcbf29ce484222325;
    for b in s.bytes() {
        h ^= b as u64;
        h = h.wrapping_mul(0x100000001b3);
    }
    h
}

/// Run the cache strategy on a stream of queries.
///
/// Returns CacheStats summarizing hit rate performance.
/// Higher hit_rate indicates a better caching strategy.
pub fn cache_strategy(queries: &[&str], capacity: u64) -> CacheStats {
    debug_assert!(capacity > 0, "Cache capacity must be positive");

    let mut cache = LruSqlCache {
        capacity,
        entries: Vec::new(),
        prefix_index: HashMap::new(),
        stats: CacheStats::default(),
        next_key: 0,
    };

    for (i, query) in queries.iter().enumerate() {
        let normalized = normalize_query(query);
        let hash = simple_hash(&normalized);

        // Exact lookup
        if let Some(pos) = cache.entries.iter().position(|(_, e)| simple_hash(&e.normalized) == hash) {
            cache.entries[pos].1.hit_count += 1;
            // Move to end (most recently used)
            let entry = cache.entries.remove(pos);
            cache.entries.push(entry);
            cache.stats.hits += 1;
            continue;
        }

        // Prefix lookup
        let prefix = extract_prefix(&normalized, 3);
        let prefix_hit = if let Some(keys) = cache.prefix_index.get(&prefix) {
            keys.iter().any(|k| cache.entries.iter().any(|(ek, _)| ek == k))
        } else {
            false
        };

        if prefix_hit {
            cache.stats.prefix_hits += 1;
            cache.stats.hits += 1;
            continue;
        }

        cache.stats.misses += 1;

        // Evict if at capacity
        if cache.entries.len() as u64 >= cache.capacity {
            if let Some((_, evicted)) = cache.entries.first() {
                let evicted_prefix = extract_prefix(&evicted.normalized, 3);
                if let Some(keys) = cache.prefix_index.get_mut(&evicted_prefix) {
                    keys.retain(|k| *k != cache.entries[0].0);
                }
            }
            cache.entries.remove(0);
            cache.stats.evictions += 1;
        }

        // Insert
        let key = cache.next_key;
        cache.next_key += 1;
        let entry = CacheEntry {
            query: query.to_string(),
            normalized: normalized.clone(),
            result_id: i as u64,
            hit_count: 0,
        };
        cache.entries.push((key, entry));

        cache
            .prefix_index
            .entry(prefix)
            .or_default()
            .push(key);
    }

    cache.stats.size = cache.entries.len() as u64;
    cache.stats
}
